normandy/recipes/models.py

```diff
@@ -174,18 +249,55 @@ class RecipeRevision(models.Model):
     name = models.CharField(max_length=255)
     action = models.ForeignKey('Action', related_name='recipe_revisions')
     arguments_json = models.TextField(default='{}', validators=[validate_json])
-    filter_expression = models.TextField(blank=False)
+    extra_filter_expression = models.TextField(blank=False)
+    channels = models.ManyToManyField(Channel)
+    countries = models.ManyToManyField(Country)
+    locales = models.ManyToManyField(Locale)
 
     @property
     def data(self):
         return {
             'name': self.name,
             'action': self.action,
             'arguments_json': self.arguments_json,
-            'filter_expression': self.filter_expression,
+            'extra_filter_expression': self.extra_filter_expression,
+            'channels': list(self.channels.all()),
+            'countries': list(self.countries.all()),
+            'locales': list(self.locales.all()),
         }
 
     @property
+    def filter_expression(self):
+        exp = ''
+
+        if self.locales.count():
+            locales = ', '.join(["'{}'".format(l.code) for l in self.locales.all()])
+            exp = 'normandy.locale in [{}]'.format(locales)
+
+        if self.countries.count():
+            countries = ', '.join(["'{}'".format(c.code) for c in self.countries.all()])
+
+            if len(exp):
+                exp += ' && '
+
+            exp += 'normandy.country in [{}]'.format(countries)
+
+        if self.channels.count():
+            channels = ', '.join(["'{}'".format(c.slug) for c in self.channels.all()])
+
+            if len(exp):
+                exp += ' && '
+
+            exp += 'normandy.channel in [{}]'.format(channels)
+
+        if self.extra_filter_expression:
+            prefix = ' && (' if len(exp) else ''
```

Comment:
I think you can just assume the statement will be wrapped in parenthesis even if it's the only thing in the expression, rather than trying to avoid them.